### PR TITLE
Allow querying multiple platforms

### PR DESCRIPTION
**Pull request by jj-style**
Fix for an old issue #45 - to allow specifying multiple `platform`s on the cli.
Behaviour is to try get the tldr page for each platform given - so if a page doesn't exist for one platform but does for another, you still get the page displayed.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -281,7 +281,10 @@ fn main() {
     }
 
     // Specify target OS
-    let platform: PlatformType = args.platform.unwrap_or_else(PlatformType::current);
+    let platforms: Vec<PlatformType> = match args.platform {
+        Some(ref p) => p.clone(),
+        None => vec![PlatformType::current()],
+    };
```
Comment (niklasmohrin) [inline src/main.rs:287]: Hm, this seems wrong. We don't need ownership of the vector, a slice should totally suffice. I guess we are running into borrowing issues later, but we should fix those instead.

Depending on how difficult this turns out to be, we can also postpone a refactor, but I would not want this kind of fixing to pile up

File: src/main.rs [hunk 2]
```diff
@@ -294,85 +297,95 @@ fn main() {
...
-    // Check cache presence and freshness
-    if !cache_updated
-        && (args.list || !args.command.is_empty())
-        && check_cache(&cache, &args, enable_styles) == CheckCacheResult::CacheMissing
-    {
-        process::exit(1);
-    }
+    for platform in platforms {
+        // Instantiate cache. This will not yet create the cache directory!
+        let cache = Cache::new(platform, &config.directories.cache_dir.path);
```
Comment (niklasmohrin) [inline src/main.rs:310]: So far, we only ever had one cache for the whole program, which makes sense, considering that we have all of the tldr-pages checked out in one directory. Right now, this code has a bug where we check whether the same directory needs to be updated (or cleared) for every platform the a user passes. I think it also lists some commands more than once, if there are multiple versions for different platforms.

I think we should stick to only having one `cache`. About adding support for multiple platforms, we could either make `Cache` have its own vector of platforms, or, just pass the platforms when they are needed. I like the latter the most, it will also lead to a solution where we don't need an owned `Vec` and don't have lifetimes on `Cache` (only in the methods that will then take `platforms: &[Platform]`).

File: src/main.rs [hunk 3]
```diff
@@ -294,85 +297,95 @@ fn main() {
...
-                .map(PathWithSource::path),
-        ) {
-            if let Err(ref e) =
-                print_page(&lookup_result, args.raw, enable_styles, args.pager, &config)
-            {
-                print_error(enable_styles, e);
-                process::exit(1);
-            }
+                .map(PathWithSource::path);
+            println!("{}", cache.list_pages(custom_pages_dir).join("\n"));
```
Comment (niklasmohrin) [inline src/main.rs:340]: I wonder if we want to display some disambiguation here: What should the output be, if we run `tldr --platform linux --platform windows --list cd`? @dbrgn 

We should check the client specification from tldr-pages/tldr for that (or ask them for a recommendation)

File: tests/lib.rs [hunk 4]
```diff
@@ -562,6 +562,34 @@ fn test_pager_flag_enable() {
...
 }
 
+#[test]
+fn test_multiple_platform_command_search() {
+    let testenv = TestEnv::new();
+    testenv.add_os_entry("linux", "linux-only", "this command only exists for linux");
+
+    testenv
+        .command()
+        .args(["--platform", "macos", "--platform", "linux", "linux-only"])
```
Comment (niklasmohrin) [inline tests/lib.rs:572]: Do we want to guarantee behavior if the page occurs for multiple platforms? Maybe try platforms in the order they were specified? In that case, we would also need a test for that

**Review by niklasmohrin — CHANGES_REQUESTED** [review]
Thank you for contributing! I have some change requests below, the main one is passing `platforms` to the methods that need it and not storing it in the `Cache`. This also includes handling multiple platforms in these methods instead of running them multiple times. Reviews might take a while here as @dbrgn and I are both relatively busy; feel free to ping us again here if we haven't checked in in a while or if you feel like we forgot you

**Comment by jj-style** [comment]
Thank you for the code review and suggestions! I agree about passing in `platform` where it's needed and refactoring it out of the `Cache`.
I'll have a crack at fixing these issues, I'm still somewhat new to rust but hopefully should be doable and will be a good learning experience!

p.s. thanks for the awesome project - it saves my bacon multiple times a day!

**Comment by niklasmohrin** [comment]
Feel free to ask about any problems you encounter, I will try to answer without unreasonable delays

**Comment by jj-style** [comment]
Happy new year! Sorry it's taken me a while to respond to your feedback. I've given it an attempt to remove the `platform` from the `Cache` and pass in `&[PlatformType]` on the methods that need it. It definitely feels better than looping the main program for each platform and creating multiple instances of the `cache`

File: src/main.rs [hunk 5]
```diff
@@ -280,8 +280,10 @@ fn main() {
         create_config_and_exit(enable_styles);
     }
 
-    // Specify target OS
-    let platform: PlatformType = args.platform.unwrap_or_else(PlatformType::current);
+    let platforms = match args.platform {
+        Some(ref platforms) => platforms.clone(),
+        None => vec![PlatformType::current()],
+    };
```
Comment (niklasmohrin) [inline src/main.rs:286]: We can avoid the clone here by borrowing into the `Option`:
```suggestion
    const FALLBACK_PLATFORMS: &'static [PlatformType] = &[PlatformType::current()];
    let platforms = args
        .platform
        .as_ref()
        .map_or(FALLBACK_PLATFORMS, Vec::as_slice);
```

File: src/main.rs [hunk 6]
```diff
@@ -295,7 +297,7 @@ fn main() {
     }
 
     // Instantiate cache. This will not yet create the cache directory!
-    let cache = Cache::new(platform, &config.directories.cache_dir.path);
+    let cache = Cache::new(config.directories.cache_dir.path.clone());
```
Comment (niklasmohrin) [inline src/main.rs:300]: Is there any particular reason for the change from `&` to `.clone`?

File: src/main.rs [hunk 7]
```diff
@@ -364,8 +370,8 @@ fn main() {
                     enable_styles,
                     &format!(
                         "Page `{}` not found in cache.\n\
-                         Try updating with `tldr --update`, or submit a pull request to:\n\
-                         https://github.com/tldr-pages/tldr",
+                            Try updating with `tldr --update`, or submit a pull request to:\n\
+                            https://github.com/tldr-pages/tldr",
```
Comment (niklasmohrin) [inline src/main.rs:374]: Please revert the changes on these lines :)

File: src/cache.rs [hunk 8]
```diff
@@ -289,10 +290,17 @@ impl Cache {
...
+        &self,
+        custom_pages_dir: Option<&Path>,
+        platforms: &[PlatformType],
+    ) -> Vec<String> {
         // Determine platforms directory and platform
         let platforms_dir = self.pages_dir().join("pages");
-        let platform_dir = self.get_platform_dir();
+        let platform_dirs: Vec<&'static str> = platforms
+            .iter()
+            .map(|&p| Cache::get_platform_dir(p))
```
Comment (niklasmohrin) [inline src/cache.rs:302]: Nit: This should use `Self` instead of `Cache` :)

**Review by niklasmohrin — COMMENTED** [review]
Thanks, looks good :)

Some of my previous comments are still open, in particular what should happen with `--list`, and whether we want to guarantee that we respect the order the platforms were given in. For the latter, I think we can always prioritize platforms that came first (as we already do) - so we would just need a test that covers this. @dbrgn Do you have opinions on behavior and whether we want to try to incorporate this into the spec?

**Comment by jj-style** [comment]
Thanks for the feedback! I have now implemented the suggestions you have made.

With regards to what should happen with `--list` when multiple platforms are provided - the current behaviour in `cache.rs` with the `pages.sort().dedup()` implies that the order the platforms are given is not important and the combined ordered set of pages should be listed.
I have added a test for this behaviour however, if we decide we want to list the pages grouped in order by the platforms supplied I can make this change

File: src/cli.rs [hunk 9]
```diff
@@ -39,9 +39,10 @@ pub(crate) struct Args {
     #[clap(
         short = 'p',
         long = "platform",
+        action = ArgAction::Append,
         possible_values = ["linux", "macos", "windows", "sunos", "osx", "android"],
     )]
-    pub platform: Option<PlatformType>,
+    pub platform: Option<Vec<PlatformType>>,
```
Comment (niklasmohrin) [inline src/cli.rs:45]: I think it should now be:
```suggestion
    pub platforms: Option<Vec<PlatformType>>,
```
Can you verify that this change doesn't change the CLI argument to `--platforms` then? I think it should work like this though

**Review by niklasmohrin — COMMENTED** [review]
Thanks, looks very good now! I have some final remarks, mostly about testing, below.

There is also [one comment from before](https://github.com/dbrgn/tealdeer/pull/300#discussion_r1008561492) that I think hasn't been addressed yet. A check should be added to make sure that when running `tldr --platform linux --platform windows xyz`, a page for linux should be preferred, even if `xyz` exists for both platforms. I think it could be tested in `test_multiple_platform_command_search`.

I thought about the behavior for `--list` some more and I think the current version is probably what we want. I am fine with the current behavior and if we ever need to change it, well, we can still change it :^)

File: src/cli.rs (hunk 9 above)
Comment (jj-style) [reply src/cli.rs]: Good spot yeah I have updated and tested this. It is still `--platform` from the `clap` `long` annotation

**Comment by jj-style** [comment]
thank you for the continuous feedback and help with this. I have pushed the changes for the testing :)

**Review by niklasmohrin — APPROVED** [review]
Looks good to me! Thank you for your contribution and sorry for the long wait :)

**Review by niklasmohrin — APPROVED** [review]
Sorry for the long wait! Your PR looks good and will be included in the next release. Thanks!

**Comment by dbrgn** [comment]
Hi, some (very late, I know) follow-up questions @niklasmohrin @jj-style:

- Was it always clear that the API should be `tldr --platform linux --platform windows` as opposed to something like `tldr --platform linux,windows`? Or was that second alternative considered?
- The [client specification](https://github.com/tldr-pages/tldr/blob/main/CLIENT-SPECIFICATION.md) does not mention the behavior in case of multiple pages. Niklas asked this in [this comment](https://github.com/dbrgn/tealdeer/pull/300#discussion_r1008560304), but I didn't see that it was addressed. Are there any other clients that support specifying multiple platforms? How does their API look like?
- (Also, the help seems confusing to me now, because it mixes a singular keyword with a plural parameter name. But I can fix that in a separate PR.)

**Comment by niklasmohrin** [comment]
- I dont remember exactly, this is just the default behavior of clap for this. I think I slightly prefer the current version, because it passes multiple platforms via multiple arguments and a parsing error could point directly to which platform is invalid. 
- I thought that the tests assert a reasonable behavior - platforms are checked in the order they are passed - but I dont fully remember and would have to dig in again. If the spec decides something else, we can change it again
- Agree, this could be improved. Maybe a note that it can be passes multiple times would be better

**Comment by jj-style** [comment]
I did not check if or how other clients support multiple platforms, sorry.
I agree with  @niklasmohrin about using the default clap behaviour but happy to see if there are other ways of doing it in clap that provide a more concise interface. 

Happy to update the help text in another PR either way!

**Comment by dbrgn** [comment]
Help text PR is here: #375
